### PR TITLE
Implement Maintenance Mode Helper 

### LINKED ISSUE #14575: Change Maintenance Mode to use TmpStore
Also check whether `\Pimcore\Tool\Admin::scheduleMaintenanceModeOnLogin()` is still relevant.

### PR DESCRIPTION
**Pull request by dvesh3**
## Changes in this pull request  
resolves #14575

### <samp>🤖 Generated by Copilot at c25a39b</samp>

This pull request replaces the deprecated maintenance mode file with a new service that uses the TmpStore model and dispatches events. It refactors the MaintenancePageListener, Bootstrap and Application classes to use the new service, and deprecates some irrelevant methods of the Admin class. It also adds the MaintenanceModeHelper and MaintenanceModeHelperInterface classes and their service definitions.

### <samp>🤖 Generated by Copilot at c25a39b</samp>

> _The `Admin` class was a mess_
> _With methods that no one would bless_
> _So they made a new helper_
> _For maintenance mode swelter_
> _And cleaned up the code with finesse_

### <samp>🤖 Generated by Copilot at c25a39b</samp>

*  Introduce a new service, `MaintenanceModeHelper`, to handle the logic for activating, deactivating and checking the maintenance mode using the `TmpStore` model instead of the maintenance mode file ([link](https://github.com/pimcore/pimcore/pull/15954/files?diff=unified&w=0#diff-f16bffa67c7a8d304f9c83fade71a9374e93647a33a12b9324f083dcc506d2c0R1-R82), [link](https://github.com/pimcore/pimcore/pull/15954/files?diff=unified&w=0#diff-d1aa6ec8c881652215889cd83455b04ac66cb54d7242cd55e97b44fa61a4fbb1R1-R26))
*  Register the `MaintenanceModeHelper` service and its interface as public services in `extensions.yaml` for dependency injection ([link](https://github.com/pimcore/pimcore/pull/15954/files?diff=unified&w=0#diff-84fbd5255e9baaaea38c0c1fcced7d47ac749c2cda4a99bf5e9688e12227c240L36-R38))
*  Replace the deprecated usage of the `Admin` class methods for maintenance mode with the `MaintenanceModeHelper` service methods in the `MaintenancePageListener`, `Bootstrap` and `Application` classes, and add deprecation notices for backward compatibility ([link](https://github.com/pimcore/pimcore/pull/15954/files?diff=unified&w=0#diff-1551473d781eb39bc192feacd11f912e168b188f3ac80eb8719d36c5b76db023R20), [link](https://github.com/pimcore/pimcore/pull/15954/files?diff=unified&w=0#diff-1551473d781eb39bc192feacd11f912e168b188f3ac80eb8719d36c5b76db023L35-R39), [link](https://github.com/pimcore/pimcore/pull/15954/files?diff=unified&w=0#diff-1551473d781eb39bc192feacd11f912e168b188f3ac80eb8719d36c5b76db023L82-R102), [link](https://github.com/pimcore/pimcore/pull/15954/files?diff=unified&w=0#diff-ae4fd3053efaec0fe50cc39c28378027e8884af72d1facbb19980ade26f36220R21-R22), [link](https://github.com/pimcore/pimcore/pull/15954/files?diff=unified&w=0#diff-ae4fd3053efaec0fe50cc39c28378027e8884af72d1facbb19980ade26f36220L74-R79), [link](https://github.com/pimcore/pimcore/pull/15954/files?diff=unified&w=0#diff-f1b341f8c516eb5173d334d53d8890aecf4038817679a3ebebd734af044c6ea1R25), [link](https://github.com/pimcore/pimcore/pull/15954/files?diff=unified&w=0#diff-f1b341f8c516eb5173d334d53d8890aecf4038817679a3ebebd734af044c6ea1L67-R96), [link](https://github.com/pimcore/pimcore/pull/15954/files?diff=unified&w=0#diff-d8b1f4ffbedfa132c950644f445ac215fb6adc74e16e35ae6f5c1a98de5a2aeaR188-R194))
*  Deprecate several unused or unrelated methods of the `Admin` class, such as `determineCsvDialect`, `getMaintenanceModeFile`, `getMaintenanceModeScheduleLoginFile`, `activateMaintenanceMode`, `deactivateMaintenanceMode`, `getMinimizedUserList`, `getMinimizedGroupList` and `getMinimizedRoleList`, and add deprecation annotations to them ([link](https://github.com/pimcore/pimcore/pull/15954/files?diff=unified&w=0#diff-d8b1f4ffbedfa132c950644f445ac215fb6adc74e16e35ae6f5c1a98de5a2aeaR123-R124), [link](https://github.com/pimcore/pimcore/pull/15954/files?diff=unified&w=0#diff-d8b1f4ffbedfa132c950644f445ac215fb6adc74e16e35ae6f5c1a98de5a2aeaR133-R135), [link](https://github.com/pimcore/pimcore/pull/15954/files?diff=unified&w=0#diff-d8b1f4ffbedfa132c950644f445ac215fb6adc74e16e35ae6f5c1a98de5a2aeaR142), [link](https://github.com/pimcore/pimcore/pull/15954/files?diff=unified&w=0#diff-d8b1f4ffbedfa132c950644f445ac215fb6adc74e16e35ae6f5c1a98de5a2aeaR166-R167), [link](https://github.com/pimcore/pimcore/pull/15954/files?diff=unified&w=0#diff-d8b1f4ffbedfa132c950644f445ac215fb6adc74e16e35ae6f5c1a98de5a2aeaR178-R179), [link](https://github.com/pimcore/pimcore/pull/15954/files?diff=unified&w=0#diff-d8b1f4ffbedfa132c950644f445ac215fb6adc74e16e35ae6f5c1a98de5a2aeaR206-R208), [link](https://github.com/pimcore/pimcore/pull/15954/files?diff=unified&w=0#diff-d8b1f4ffbedfa132c950644f445ac215fb6adc74e16e35ae6f5c1a98de5a2aeaR225-R227), [link](https://github.com/pimcore/pimcore/pull/15954/files?diff=unified&w=0#diff-d8b1f4ffbedfa132c950644f445ac215fb6adc74e16e35ae6f5c1a98de5a2aeaR239-R241))

### CODE REVIEW

File: lib/Console/Application.php [hunk 1]
```diff
@@ -64,31 +65,42 @@ public function __construct(KernelInterface $kernel)
...
-            if ($event->getCommand() instanceof DoctrineCommand && $prefix = $event->getInput()->getOption('prefix')) {
-                $kernel->getContainer()->get(FilteredMigrationsRepository::class)->setPrefix($prefix);
-                $kernel->getContainer()->get(FilteredTableMetadataStorage::class)->setPrefix($prefix);
-            }
+        $maintenanceModeHelper = $kernel->getContainer()->get(MaintenanceModeHelperInterface::class);
+        $dispatcher->addListener(ConsoleEvents::COMMAND, function (ConsoleCommandEvent $event)
+            use ($kernel, $maintenanceModeHelper) {
+                // skip if maintenance mode is on and the flag is not set
+                if ($maintenanceModeHelper->isActive() || Admin::isInMaintenanceMode() &&
+                    !$event->getInput()->getOption('ignore-maintenance-mode')
```
Comment (kingjia90) [inline lib/Console/Application.php]:
```suggestion
                if (($maintenanceModeHelper->isActive() || Admin::isInMaintenanceMode()) &&
                    !$event->getInput()->getOption('ignore-maintenance-mode')
```
similarly to https://github.com/pimcore/pimcore/pull/15954/files?w=1#diff-ae4fd3053efaec0fe50cc39c28378027e8884af72d1facbb19980ade26f36220R78

File: lib/Tool/Admin.php [hunk 2]
```diff
@@ -157,6 +163,8 @@ public static function activateMaintenanceMode(?string $sessionId): void
     }
 
     /**
+     * @deprecated Use MaintenanceModeHelperInterface::activate instead.
```
Comment (kingjia90) [inline lib/Tool/Admin.php]:
```suggestion
     * @deprecated Use MaintenanceModeHelperInterface::deactivate instead.
```

File: lib/Tool/MaintenanceModeHelper.php [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+        \Pimcore::getEventDispatcher()->dispatch(new GenericEvent(), SystemEvents::MAINTENANCE_MODE_DEACTIVATE);
+    }
+
+    public function isActive(string $matchSessionId = null): bool
+    {
+        if ($maintenanceModeEntry = $this->getEntry()) {
+            if ($matchSessionId && $matchSessionId !== $maintenanceModeEntry) {
+                return false;
+            }
+            return true;
```
Comment (kingjia90) [inline lib/Tool/MaintenanceModeHelper.php]: the logic here was inverted, if the session id matches, for me maintenance mode should be not active (false)

```suggestion
            if ($matchSessionId && $matchSessionId === $maintenanceModeEntry) {
                return false;
            }
            return true;
```
or [invert the return ](https://github.com/pimcore/pimcore/pull/15954/files/49112a8d37431842410ded9fd4954ec6218ee38d?w=1#r1330470184)

File: lib/Tool/Admin.php [hunk 4]
```diff
@@ -202,6 +222,9 @@ public static function isMaintenanceModeScheduledForLogin(): bool
         return false;
     }
 
+    /**
+     * @deprecated
+     */
     public static function scheduleMaintenanceModeOnLogin(): void
```
Comment (kingjia90) [inline lib/Tool/Admin.php:228]: The replacement for `scheduleMaintenanceModeOnLogin` might be missing. 
See also https://github.com/pimcore/admin-ui-classic-bundle/pull/274#pullrequestreview-1633791218

**Review by kingjia90 — CHANGES_REQUESTED** [review]
Found some minor issues but overall LGTM
- [x] upgrade note entry about deprecations and replacement
- [x] bc layer to the console command `bin/console pimcore:maintenance-mode`
https://github.com/pimcore/pimcore/blob/49112a8d37431842410ded9fd4954ec6218ee38d/bundles/CoreBundle/src/Command/MaintenanceModeCommand.php#L58

File: lib/Tool/MaintenanceModeHelper.php [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+        }
+
+        return false;
```
Comment (kingjia90) [inline lib/Tool/MaintenanceModeHelper.php]: this way would have one return less
```suggestion
        if ($maintenanceModeEntry = $this->getEntry()) {
            if ($matchSessionId && $matchSessionId !== $maintenanceModeEntry) {
                return true;
            }
        }
        return false;
```

File: lib/Tool/Admin.php (hunk 4 above)
Comment (dvesh3) [reply lib/Tool/Admin.php:228]: there's no replacement as it will be removed in Pimcore 12

File: lib/Tool/MaintenanceModeHelper.php (hunk 3 above)
Comment (dvesh3) [reply lib/Tool/MaintenanceModeHelper.php]: Fixed with merging the or suggestion

File: lib/Tool/Admin.php (hunk 4 above)
Comment (kingjia90) [reply lib/Tool/Admin.php:228]: then suggesting to add the removal note on the `@deprecated` tag line

**Comment by dvesh3** [comment]
> Found some minor issues but overall LGTM
> 
> * [x]  upgrade note entry about deprecations and replacement
> * [x]  bc layer to the console command `bin/console pimcore:maintenance-mode`
>   https://github.com/pimcore/pimcore/blob/49112a8d37431842410ded9fd4954ec6218ee38d/bundles/CoreBundle/src/Command/MaintenanceModeCommand.php#L58

@kingjia90 Done

**Comment by dvesh3** [comment]
> Found some minor issues but overall LGTM
> 
> * [x]  upgrade note entry about deprecations and replacement
> * [x]  bc layer to the console command `bin/console pimcore:maintenance-mode`
>   https://github.com/pimcore/pimcore/blob/49112a8d37431842410ded9fd4954ec6218ee38d/bundles/CoreBundle/src/Command/MaintenanceModeCommand.php#L58

@kingjia90 Done

**Pull request merged**